#### First we need to import xome library for our analysis.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans


import folium # map rendering library

print('Libraries imported.')

/usr/lib/python3/dist-packages/requests/__init__.py:80: RequestsDependencyWarning: urllib3 (1.25.8) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)


Libraries imported.


### Part1: Background of the data
##### We want to look at the boroughs of Totonto in Canada,by combing two csv together, we get the dataframe.

In [2]:
table = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
df=table[0]
dfd1=df[df.Borough!="Not assigned"]
df = dfd1.reset_index(drop=True)
geo=pd.read_csv("http://cocl.us/Geospatial_data")
k=df["Postal Code"]
pc=list(k.values)
lat=[]
lon=[]
for n in pc:
    lat.append(float(geo[geo["Postal Code"]==n]["Latitude"].values))
    lon.append(float(geo[geo["Postal Code"]==n]["Longitude"].values))
df["Latitude"]=lat
df["Longitude"]=lon
neighborhoods=df

In [3]:
neighborhoods

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

In [4]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [5]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))


location 

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Location(Toronto, Golden Horseshoe, Ontario, M5H 2N2, Canada, (43.6534817, -79.3839347, 0.0))

In [6]:

# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='orange',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

##### So here is the map of Toronto, it has 10 boroughs and 103 neighborhoods.
##### The geograpical coordinate of Toronto are 43.6534817, -79.3839347.

### Part 2: A description of the data and how it will be used to solve the problem.

##### If a person in Tonronto would love to move in the city becasue of the changes of job, 
##### He still wants to move to the same environment like the plaves he used to live in.

In [7]:
allb=list(set(list(neighborhoods['Borough'])))

##### So we first assume the persone lives in the borough of Downtown Toronto

In [8]:
dt_data = neighborhoods[neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
dt_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

In [9]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Tonronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Tonronto are 43.6563221, -79.3809161.


##### Now we need to use the Foursquare to explore the neighborhood

In [10]:
CLIENT_ID = 'JRFGNARG43DGGKORVKO3DVU5ADFYSFZLKK2A5DWWWBQHDY42' # your Foursquare ID
CLIENT_SECRET = 'OKZFAF3MRV0UCG2NATSBMUHWFPY35NYGFYOI1JOBV5VT0H5G' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: JRFGNARG43DGGKORVKO3DVU5ADFYSFZLKK2A5DWWWBQHDY42
CLIENT_SECRET:OKZFAF3MRV0UCG2NATSBMUHWFPY35NYGFYOI1JOBV5VT0H5G


In [11]:
LIMIT=100
radius=500
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [12]:
dt_venues = getNearbyVenues(names=dt_data['Neighborhood'],
                                   latitudes=dt_data['Latitude'],
                                   longitudes=dt_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [13]:
dt_venues.groupby('Neighborhood').count()
print('There are {} uniques categories.'.format(len(dt_venues['Venue Category'].unique())))

There are 208 uniques categories.


##### We got there are 208 uniques categories in Downtown Toronto.

In [14]:
# one hot encoding
dt_onehot = pd.get_dummies(dt_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
dt_onehot['Neighborhood'] = dt_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [dt_onehot.columns[-1]] + list(dt_onehot.columns[:-1])
dt_onehot = dt_onehot[fixed_columns]

dt_grouped = dt_onehot.groupby('Neighborhood').mean().reset_index()

In [15]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [16]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = dt_grouped['Neighborhood']

for ind in np.arange(dt_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dt_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant              Beer Bar   
1        Airport Lounge      Airport Terminal                 Plane   
2        Sandwich Place    Italian Restaurant                  Café   
3                  Café                  Park            Baby Store   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery            Restaurant           Cheese Shop   
1           Coffee Shop      Sculpture Garden         Boat or Ferry   
2   Japanese Restaurant       Bubble Tea Shop      Department Store   
3    Italian Restaurant                 Diner             Nightclub   
4               Gay Bar                 Hotel                   Pub   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0                  Café   Japanese Restaurant  Eastern European Restaurant  
1   Rental Car Location       Harbor / Marina                     Boutique  
2           Salad Place       Thai Restaurant                 Burger Joint  
3           Candy Store            Restaurant                  Coffee Shop  
4       Bubble Tea Shop           Men's Store     Mediterranean Restaurant

In [17]:
# set number of clusters
kclusters = 5

dt_grouped_clustering = dt_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 0, 4, 3, 2, 2, 2, 2, 4, 2], dtype=int32)

In [18]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

dt_merged = dt_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
dt_merged = dt_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

dt_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               4           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               2        Clothing Store   
3  43.651494 -79.375418               2                  Café   
4  43.644771 -79.373306               2           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                        Pub                  Park   
1                 Diner           Sushi Restaurant           Yoga Studio   
2           Coffee Shop  Middle Eastern Restaurant                  Café   
3           Coffee Shop                 Restaurant   American Restaurant   
4          Cocktail Bar         Seafood Restaurant              Beer Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                  Café               Theater   
1   Arts & Crafts Store        Sandwich Place        Discount Store   
2       Bubble Tea Shop   Japanese Restaurant        Cosmetics Shop   
3             Gastropub          Cocktail Bar        Clothing Store   
4                Bakery            Restaurant           Cheese Shop   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0           Yoga Studio          Dessert Shop                   Restaurant  
1              Beer Bar    Italian Restaurant                Burrito Place  
2                 Diner    Italian Restaurant               Lingerie Store  
3   Moroccan Restaurant        Cosmetics Shop                     Creperie  
4                  Café   Japanese Restaurant  Eastern European Restaurant

In [19]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dt_merged['Latitude'], dt_merged['Longitude'], dt_merged['Neighborhood'], dt_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### The map above shows the clusters in the Downtown Tonronto, so if the person wants to move in the same borough, it would be great to choose the places with the same color in this map.

## part3: More solutions
#### If the person needs to move between differnt boroughs.

In [20]:
yk_data = neighborhoods[neighborhoods['Borough'] == "York"].reset_index(drop=True)
yk_venues = getNearbyVenues(names=yk_data['Neighborhood'],
                                   latitudes=yk_data['Latitude'],
                                   longitudes=yk_data['Longitude']
                                  )
yk_venues

Humewood-Cedarvale
Caledonia-Fairbanks
Del Ray, Mount Dennis, Keelsdale and Silverthorn
Runnymede, The Junction North
Weston


Neighborhood  Neighborhood Latitude  \
0                                 Humewood-Cedarvale              43.693781   
1                                 Humewood-Cedarvale              43.693781   
2                                 Humewood-Cedarvale              43.693781   
3                                 Humewood-Cedarvale              43.693781   
4                                Caledonia-Fairbanks              43.689026   
5                                Caledonia-Fairbanks              43.689026   
6                                Caledonia-Fairbanks              43.689026   
7                                Caledonia-Fairbanks              43.689026   
8   Del Ray, Mount Dennis, Keelsdale and Silverthorn              43.691116   
9   Del Ray, Mount Dennis, Keelsdale and Silverthorn              43.691116   
10  Del Ray, Mount Dennis, Keelsdale and Silverthorn              43.691116   
11  Del Ray, Mount Dennis, Keelsdale and Silverthorn              43.691116   
12  Del Ray, Mount Dennis, Keelsdale and Silverthorn              43.691116   
13                     Runnymede, The Junction North              43.673185   
14                     Runnymede, The Junction North              43.673185   
15                     Runnymede, The Junction North              43.673185   
16                     Runnymede, The Junction North              43.673185   
17                                            Weston              43.706876   
18                                            Weston              43.706876   

    Neighborhood Longitude                      Venue  Venue Latitude  \
0               -79.428191             Cedarvale Park       43.692535   
1               -79.428191           Cedarvale Ravine       43.690188   
2               -79.428191            Glen Cedar Park       43.695399   
3               -79.428191           Phil White Arena       43.691303   
4               -79.453512                 Nairn Park       43.690654   
5               -79.453512              Maximum Woman       43.690651   
6               -79.453512             Fairbanks Pool       43.691959   
7               -79.453512     Fairbank Memorial Park       43.692028   
8               -79.476013                     Subway       43.690218   
9               -79.476013                Dollar Tree       43.690296   
10              -79.476013                 McDonald's       43.692926   
11              -79.476013             Z Bar & Grille       43.690673   
12              -79.476013     New York Fried Chicken       43.688473   
13              -79.487262            195 Jane Rocket       43.672335   
14              -79.487262          High Park Brewery       43.669903   
15              -79.487262          Welcome Food Mart       43.669701   
16              -79.487262            Starving Artist       43.669860   
17              -79.518188                   Elm Park       43.706139   
18              -79.518188  Olympic convenience store       43.704486   

    Venue Longitude        Venue Category  
0        -79.428705                 Field  
1        -79.426106                 Trail  
2        -79.429253            Playground  
3        -79.431761          Hockey Arena  
4        -79.456300                  Park  
5        -79.456333         Women's Store  
6        -79.448922                  Pool  
7        -79.448924                  Park  
8        -79.474050        Sandwich Place  
9        -79.474667        Discount Store  
10       -79.479982  Fast Food Restaurant  
11       -79.472764                   Bar  
12       -79.480948  Fast Food Restaurant  
13       -79.492634              Bus Line  
14       -79.483430               Brewery  
15       -79.483491         Grocery Store  
16       -79.483110        Breakfast Spot  
17       -79.515869                  Park  
18       -79.515789     Convenience Store

In [21]:
combine=[yk_data,dt_data]
result_data = pd.concat(combine)

In [22]:
# one hot encoding
yk_onehot = pd.get_dummies(yk_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
yk_onehot['Neighborhood'] = yk_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [yk_onehot.columns[-1]] + list(yk_onehot.columns[:-1])
yk_onehot = yk_onehot[fixed_columns]

yk_grouped = yk_onehot.groupby('Neighborhood').mean().reset_index()


In [23]:
merge=[dt_grouped,yk_grouped]
result = pd.concat(merge)
result=result.fillna(0)

In [24]:
result

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.015385   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.025974   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.030303   
11                          Regent Park, Harbourfront     0.022222   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010309   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.028571   
0                                 Caledonia-Fairbanks     0.000000   
1    Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
2                                  Humewood-Cedarvale     0.000000   
3                       Runnymede, The Junction North     0.000000   
4                                              Weston     0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.058824            0.058824      0.058824   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.012987  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000   
17           0.000000  0.000000            0.000000      0.000000   
18           0.000000  0.000000            0.000000      0.000000   
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.000000            0.000000      0.000000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0         0.000000         0.000000          0.000000             0.000000   
1         0.117647         0.176471          0.117647             0.000000   
2         0.000000         0.000000          0.000000             0.000000   
3         0.000000         0.000000          0.000000             0.000000   
4         0.000000         0.000000          0.000000             0.012987   
5         0.000000         0.000000          0.000000             0.040000   
6         0.000000

In [25]:
kclusters = 5

result_clustering = result.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(result_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
result_venues_sorted = pd.DataFrame(columns=columns)
result_venues_sorted['Neighborhood'] = result['Neighborhood']

for ind in np.arange(dt_grouped.shape[0]):
    result_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dt_grouped.iloc[ind, :], num_top_venues)

result_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant              Beer Bar   
1        Airport Lounge      Airport Terminal                 Plane   
2        Sandwich Place    Italian Restaurant                  Café   
3                  Café                  Park            Baby Store   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery            Restaurant           Cheese Shop   
1           Coffee Shop      Sculpture Garden         Boat or Ferry   
2   Japanese Restaurant       Bubble Tea Shop      Department Store   
3    Italian Restaurant                 Diner             Nightclub   
4               Gay Bar                 Hotel                   Pub   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0                  Café   Japanese Restaurant  Eastern European Restaurant  
1   Rental Car Location       Harbor / Marina                     Boutique  
2           Salad Place       Thai Restaurant                 Burger Joint  
3           Candy Store            Restaurant                  Coffee Shop  
4       Bubble Tea Shop           Men's Store     Mediterranean Restaurant

In [27]:
# add clustering labels
result_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

result_merged = result_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
result_merged = result_merged.join(result_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

result_merged # check the last columns!


Postal Code           Borough  \
0          M6C              York   
1          M6E              York   
2          M6M              York   
3          M6N              York   
4          M9N              York   
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M5E  Downtown Toronto   
5          M5G  Downtown Toronto   
6          M6G  Downtown Toronto   
7          M5H  Downtown Toronto   
8          M5J  Downtown Toronto   
9          M5K  Downtown Toronto   
10         M5L  Downtown Toronto   
11         M5S  Downtown Toronto   
12         M5T  Downtown Toronto   
13         M5V  Downtown Toronto   
14         M4W  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M4X  Downtown Toronto   
17         M5X  Downtown Toronto   
18         M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                  Humewood-Cedarvale  43.693781 -79.428191   
1                                 Caledonia-Fairbanks  43.689026 -79.453512   
2    Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.691116 -79.476013   
3                       Runnymede, The Junction North  43.673185 -79.487262   
4                                              Weston  43.706876 -79.518188   
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         Berczy Park  43.644771 -79.373306   
5                                  Central Bay Street  43.657952 -79.387383   
6                                            Christie  43.669542 -79.422564   
7                            Richmond, Adelaide, King  43.650571 -79.384568   
8   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
9            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
10                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
11                     University of Toronto, Harbord  43.662696 -79.400049   
12          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
13  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
14                                           Rosedale  43.679563 -79.377529   
15                                     Stn A PO Boxes  43.646435 -79.374846   
16                        St. James Town, Cabbagetown  43.667967 -79.367675   
17             First Canadian Place, Underground city  43.648429 -79.382280   
18                               Church and Wellesley  43.665860 -79.383160   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                4                   NaN                   NaN   
1                3                   NaN                   NaN   
2                2                   NaN                   NaN   
3                0                   NaN                   NaN   
4                1                   NaN                   NaN   
0                0           Coffee Shop                Bakery   
1                0           Coffee Shop                 Diner   
2                0        Clothing Store           Coffee Shop   
3                0                  Café           Coffee Shop   
4                0           Coffee Shop          Cocktail Bar   
5                0           Coffee Shop        Sandwich Place   
6                0         Grocery Store                  Café   
7                0           Coffee Shop            Restaurant   
8                0           Coffee Shop              Aquarium   
9                0           Coffee Shop                 Hotel   
10               0           Coffee Shop            Restaurant   
11  

In [29]:
# create map
result_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(result_merged['Latitude'], result_merged['Longitude'], result_merged['Neighborhood'], result_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(result_clusters)
       
result_clusters

#### We use the similiar strategies to get the map above the give the person a better idea on how to move with similiar environment.

## In this way, we can give the person advice on how to move to enjoy a similiar environment.